services/inputhost/pathCache.go

```diff
@@ -508,3 +536,29 @@ func (pathCache *inPathCache) getState() *admin.DestinationState {
 
 	return destState
 }
+
+func (pathCache *inPathCache) drainExtent(extUUID string, updateUUID string, drainWG *sync.WaitGroup) {
+	defer drainWG.Done()
+	pathCache.RLock()
+
+	extCache, ok := pathCache.extentCache[extentUUID(extUUID)]
+	if !ok {
+		// draining called for extent which is not present
+		// just return
+		pathCache.RUnlock()
+		return
+	}
+
+	if extCache.connection.prepForClose() {
+		var connDrainWG sync.WaitGroup
+		// first send DRAIN command to all connections
+		// then start draining extents
+		pathCache.drainConnections(updateUUID, &connDrainWG)
+		// while drain no need to hold the lock to make sure we
+		// can proceed forward
+		pathCache.RUnlock()
+		// do best effort waiting to notify all connections here
+		common.AwaitWaitGroup(&connDrainWG, connWGTimeout)
+		extCache.connection.stopWrite()
+	}
```

Comment:
if prepForClose() returns false, which means it the extent is already either closed or being closed, then we just leave the pathCache RLock hanging.. it must be unlocked here (pathCache.RUnlock()).. 
@GuillaumeBailey @thuningxu